hpcbench/benchmark/custream.py

```diff
@@ -0,0 +1,176 @@
+                CUDAStreamExtractor._parse_line(line, metrics)
+        return metrics
+
+    @classmethod
+    def _parse_line(cls, line, metrics):
+        for sect, regex in cls.REGEX.items():
+            search = regex.search(line)
+            if search:
+                metrics[sect + "_bandwidth"] = float(search.group(1))
+                metrics[sect + "_avg_time"] = float(search.group(2))
+                metrics[sect + "_min_time"] = float(search.group(3))
+                metrics[sect + "_max_time"] = float(search.group(4))
+                return
+
+
+class CUDAStream(Benchmark):
+    """Benchmark wrapper for the cuda-stream utility
+    """
+    name = 'custream'
+
+    description = "Provides memory bandwidth benchmarking for NVIDIA GPUs."
+
+    DEFAULT_EXECUTABLE = 'stream'
+    DEFAULT_THREADS_PER_BLOCK = [64, 128, 192, 256, 512, 1024]
+
+    def __init__(self):
+        super(CUDAStream, self).__init__(
+            attributes=dict(
+                blocksizes=CUDAStream.DEFAULT_THREADS_PER_BLOCK,
+                executable=CUDAStream.DEFAULT_EXECUTABLE,
+            )
+        )
+
+    @cached_property
+    def executable(self):
+        """Get path to iperf executable
+        """
+        return self.attributes['executable']
+
+    @property
+    def blocksizes(self):
+        """List of threads per block the command is executed with"""
+        return [
+            str(e)
+            for e in self.attributes['blocksizes']
+        ]
+
+    def execution_matrix(self, context):
+        del context  # unused
+        for nthreads in self.blocksizes:
+            yield dict(
+                category=CUDAStream.name,
+                command=[find_executable(self.executable),
+                         '-b {threads}'.format(threads=nthreads)
+                         ],
+                metas=dict(blocksizes=nthreads),
+                environment=dict(),
+            )
+
+    @cached_property
+    def metrics_extractors(self):
+        return CUDAStreamExtractor()
+
+    @property
+    def plots(self):
+        return {
+            CUDAStream.name: [
+                dict(
+                    name="{hostname} {category} bandwidth",
+                    series=dict(
+                        metas=['blocksizes'],
+                        metrics=[
+                            'copy_min_time', 'copy_avg_time',
+                            'copy_max_time', 'scale_min_time',
+                            'scale_avg_time', 'scale_max_time',
+                            'add_min_time', 'add_avg_time',
+                            'add_max_time', 'triad_min_time',
+                            'triad_avg_time', 'triad_max_time',
+                        ],
+
+                    ),
+                    plotter=CUDAStream.plot_bandwidth
+                ),
+            ]
+        }
+
+    @classmethod
+    def plot_bandwidth(cls, plt, description, metas, metrics):
+        """Generate timings plot
```

Comment:
Have you tested it manually? To do:

1. write a YAML file launching this benchmark
1. execute the benchmark with `ben-sh`
1. generate the plots with `ben-plop <REPORT_PATH>`

-> REPORT_PATH should then contains some pictures.

Reply:
tested! and in the process i found a few more bugs in my code - thanks!